proposed/2023/release-staging-and-deprecation.md

```diff
@@ -0,0 +1,199 @@
+
+## Summary
+
+This proposal introduces a concept for grouping packages that are published to NuGet.
+This will allow one to easily identify all the packages released as part of a single group as well as request bulk-processing
+ for all packages which belong to the same group. For example, all packages published as part of .NET 8 release will be staged
+ together and then also can be deprecated together by a single call to NuGet APIs.
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+Many products (Apps / Tools / Frameworks / Utilities / etc. ...) consist of multiple interdependent packages which should be published together as a single set.
+ Currently, NuGet provides no functionality for definding / grouping packages into such sets.
+ This results in package owners having to keep track of these package sets on their own.
+ There are three specific sets of scenarios which this feature will simplify dramatically:
+1. When preparing a release, product owners would like to have a mechanism to pre-stage their product packages,
+ so that the release will consist only of a single action of marking the set of packages as `published`,
+ which would result all the packages in the set to be marked as published at the same time.
+2. Sometimes developers have to take a dependency on several related packages in their projects, which are part of the same "release".
+ Without having a concept of the `release` developers have to guess which versions of those packages are compatible.
+ As a result, most of the time developers would simply pick the latest version of all these packages, which sometimes can lead to incompatibility issues down the road.
+3. Keeping a package healthy on NuGet also means keeping the list of packages aligned with it support policy.
+Hence, package owners try to deprecate packages which are out of support. Without ability to group packages,
+ package owners end up keeping track of all the packages which belong to the same release and run the deprecation over a discrete set of packages when necessary. Being able to group packages together would enable deprecating a whole release at once, knowing that all the packages which belong to that release will get deprecated.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+A new NuGet CLI command `stage` will be added, which will allow a user to stage a package to nuget, specifying a `group` parameter.
+ Staged packages will be uploaded to NuGet and will be validated during the staging process, so that the staging fails
+ if validation fails for a package in a group. After uploading multiple packages, the user user can then call `nuget publish [group]`
```

Comment:
How would publisher know if some packages failed tests and hence need to be re-staged? Is this command going to be blocking? Do we need a command option to get status? Or will there be a notification (email or otherwise)?

Reply:
The validation is going to happen during the staging phase. And I have a whole section (Asynchronous behavior) about the ned for status reporting and the need for a operation status command (line 126)

Reply:
Sorry, I didn't mean to introduce a new way for pushing. I meant to still use the `nuget push` command here, just with an additional `group` parameter.